### PR TITLE
feat: support kotlin, jetpack compose, TxRxChecker

### PR DESCRIPTION
**Pull request by yeonns**
## Summary

Euphony will be used easily if it supports android compose.
So, it is a starting point - implemented ```TxRxChecker```

```TxRxChecker``` has 2 components
- input text field (it will generate random string, default string length is 3)
- play button (if you click this button, euphony Tx/Rx will work)
- Display result at input text field
  - Success: background will be changed to LightGreen
  - Fail: background will be changed to LightRed

### Initial State
스크린샷, 2022-08-25 21-06-22

### Success Result
스크린샷, 2022-08-25 21-06-40

### Fail Result
스크린샷, 2022-08-25 21-06-32

## Build & Test Result
스크린샷, 2022-08-21 00-03-54

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 buildscript {
+    ext.kotlin_version = '1.7.10'
```
Comment (designe) [inline build.gradle]: Hot new version!

File: euphony/src/main/java/co/euphony/util/EuCodec.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+		Arrays.fill(base40Index, (-1).toByte())
+		for (i in Base40.indices)
+			base40Index[Base40[i].code] = i.toByte()
+	}
+
+	companion object {
+		private const val BASE40 = 40
+		protected var base40Index = ByteArray(256)
+	}
+}
```
Comment (designe) [inline euphony/src/main/java/co/euphony/util/EuCodec.kt:20]: Great migration java to kt for EuCodec 👍

File: build.gradle [hunk 3]
```diff
@@ -1,6 +1,9 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 buildscript {
-    ext.kotlin_version = '1.7.10'
+    ext {
+        kotlin_version = '1.7.0'
```
Comment (designe) [inline build.gradle:4]: Is there some reason `kotlin_version`'s downgrade?

File: build.gradle [hunk 4]
```diff
@@ -9,7 +12,7 @@ buildscript {
         }
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:4.2.2'
+        classpath 'com.android.tools.build:gradle:7.2.2'
```
Comment (designe) [inline build.gradle:15]: `Euphony` used too old version of gradle version. haha
Thanks to update 👍

File: gradle.properties [hunk 5]
```diff
@@ -1 +1,2 @@
-android.useAndroidX = true
\ No newline at end of file
+android.useAndroidX = true
+org.gradle.jvmargs=-Xmx2048m
```
Comment (designe) [inline gradle.properties:2]: Good to setting daemon heap size up from 1024m :)
```
Default value : -Xmx1024m -XX : MaxPermSize = 256M
```

File: euphony/src/test/java/euphony/lib/ui/viewmodel/TxRxCheckerViewModelTest.kt [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        Dispatchers.setMain(mainThreadSurrogate)
+    }
+
+    @After
+    fun tearDown() {
+        Dispatchers.resetMain()
+    }
+
+    @Test
+    fun `if start, livedata should be initialized and Euphony Tx & Rx should work`() {
```
Comment (designe) [inline euphony/src/test/java/euphony/lib/ui/viewmodel/TxRxCheckerViewModelTest.kt:55]: I really like test functions' name :)
I didn't know express like this using `` 👍

**Review by designe — COMMENTED** [review]
That's good news to support kotlin with java & composable UI component!

I think that `being able to use ui components directly from the euphony library` will bring about an amazing change of experience.

I really appreciate your contribution! 👍 
This PR seems to be important enough to bring about a version change.
So, I need to have more time to review code. 

Let me finish the review as soon as possible!

File: build.gradle (hunk 4 above)
Comment (yeonns) [reply build.gradle:15]: To enable ```compose```, gradle version update was needed :) Thank you!

File: build.gradle (hunk 3 above)
Comment (yeonns) [reply build.gradle:4]: - compose ```v1.2.0``` -> kotlin ```1.7.0``` required
- compose ```v1.2.1``` -> kotlin ```1.7.10``` required

When I tried compose ```v.1.2.1``` which is the latest version, but it fails to build euphony, so I downgraded version of compose and kotlin 😢 
Now, I'm testing again whether compose ```v1.2.1```. If it is working, I will upgrade the version and let you know :)

**Comment by yeonns** [comment]
@designe I updated jvmargs and compose version :smile: Including this commit, please review PR more :)

File: euphony/src/main/java/co/euphony/util/Constants.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package co.euphony.util
+
+object Constants {
+
+    // TxRxChecker Test Tags
+    const val TAG_INPUT = "TxRxCheckerInput"
+    const val TAG_BUTTON = "TxRxCheckerButton"
+    const val TAG_RESULT_SUCCESS = "TxRxCheckerResultSuccess"
+    const val TAG_RESULT_FAIL = "TxRxCheckerResultFail"
+}
```
Comment (designe) [inline euphony/src/main/java/co/euphony/util/Constants.kt:10]: We already have Constants class in `co.euphony.common` package.
So, those string constants should be merged with `co.euphony.common.Constants`

**Review by designe — COMMENTED** [review]
As the gradle version is updated to 7.2.2 through this patch, an issue occurs where gtest results are not displayed properly.

Last year, we resolved related issue through deleting gradle user's cache.
refer to : https://github.com/euphony-io/euphony/issues/46

To solve this, modify build.gradle as shown below, and it works normally. 
It seems necessary to check that this does not cause additional issues.

`build.gradle` in root
```groovy
task cleanCache(type: Delete) {
    def cppBuildDir = "euphony/.cxx/Debug"
    delete fileTree(cppBuildDir) {
        include("**/CMakeSystem.cmake")
    }

    doLast {
        println "Clear gradle cache."
    }
}
```

`build.gradle` (:euphony)
```groovy
gradle.buildFinished {
        def abiArr = ['arm64-v8a', 'armeabi-v7a', 'x86', 'x86_64']
        def logfile = "android_gradle_build_stdout_targets.txt"
        def lastWordIfThereAreNoDevice = "Linking"
        def resultStr = "\n"
        def noDevice = true
        for (abi in abiArr) {
            def filePath = "euphony/.cxx/Debug/555j172j/$abi/$logfile" // here should be modified.
            def file = new File(filePath)

            if (file.exists()) {
                def lines = file.readLines()
                if (lines.get(lines.size() - 1).contains(lastWordIfThereAreNoDevice))
                    continue
                noDevice = false
                resultStr += "> $abi/$logfile\n"
                resultStr += lines.join("\n")
            }
        }
        if (noDevice)
            println "The build was successful, not verified by unit tests. For unit testing, connect a virtual device or Android device to adb."
        else
            println "Print build_stdout $resultStr"
    }
```

Let me analyze code deeply :)

**Comment by yeonns** [comment]
@designe 
As you suggested, I updated ```build.gradle``` and moved constants :smile: 
Thank you for detailed reviews!

**Comment by designe** [comment]
@yeonns Great :) Thanks for reflecting my suggests.
Finally, we need to resolve the below issue.
```
 Process: com.example.myapplication, PID: 22298
    java.lang.NoClassDefFoundError: Failed resolution of: Landroidx/compose/runtime/livedata/LiveDataAdapterKt;
        at co.euphony.ui.TxRxCheckerKt.TxRxChecker(TxRxChecker.kt:37)
```

After porting to sample app, above issue occurred.
I'm gonna check why this issue happened :)

**Comment by yeonns** [comment]
@designe Now, I changed from ```LiveData``` to ```StateFlow```. So, user can only import euphony dependencies and just call ```TxRxChecker()```!

File: build.gradle [hunk 8]
```diff
@@ -1,8 +1,8 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 buildscript {
     ext {
-        kotlin_version = '1.7.10'
-        compose_version = '1.2.1'
+        kotlin_version = '1.7.0'
+        compose_version = '1.2.0'
```
Comment (designe) [inline build.gradle:5]: Thanks for change kotlin & compose version :)

File: euphony/build.gradle [hunk 9]
```diff
@@ -59,7 +59,7 @@ android {
     }
 
     composeOptions {
-        kotlinCompilerExtensionVersion "1.3.0-rc02"
+        kotlinCompilerExtensionVersion compose_version
```
Comment (designe) [inline euphony/build.gradle:62]: :+1:

File: euphony/src/androidTest/java/co/euphony/ui/TxRxCheckerImplTest.kt [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        imageBitmap.readPixels(buffer, 0, 0, imageBitmap.width - 1, imageBitmap.height - 1)
+        val pixelColors = PixelMap(buffer, 0, 0, imageBitmap.width - 1, imageBitmap.height - 1)
+
+        (0 until imageBitmap.width).forEach { x ->
+            (0 until imageBitmap.height).forEach { y ->
+                if (pixelColors[x, y] == tint) return this
+            }
+        }
+        throw AssertionError("Assert failed: The component does not contain the color")
+    }
```
Comment (designe) [inline euphony/src/androidTest/java/co/euphony/ui/TxRxCheckerImplTest.kt:106]: Super interesting background check!

File: euphony/src/main/java/co/euphony/common/Constants.java [hunk 11]
```diff
@@ -24,8 +24,8 @@ public class Constants {
 	// TxRxChecker Test Tags
 	public static final String TAG_INPUT = "TxRxCheckerInput";
 	public static final String TAG_BUTTON = "TxRxCheckerButton";
-	public static final String TAG_RESULT_SUCCESS = "TxRxCheckerResultSuccess";
-	public static final String TAG_RESULT_FAIL = "TxRxCheckerResultFail";
+	public static final String PLAY_BUTTON = "▶";
+	public static final String PROGRESS_BUTTON = "▷";
```
Comment (designe) [inline euphony/src/main/java/co/euphony/common/Constants.java:28]: More Simple :)

**Review by designe — APPROVED** [review]
Wow, That's really great job!

File: euphony/build.gradle [hunk 12]
```diff
@@ -45,38 +46,55 @@ android {
...
-        }
+        })
+
         if (noDevice)
-            println "The build was successful, not verified by unit tests. For unit testing, connect a virtual device or Android device to adb."
+            println "\nThe build was successful, not verified by unit tests. For unit testing, connect a virtual device or Android device to adb."
         else
-            println "Print build_stdout $resultStr"
+            println "\nPrint build_stdout $resultStr"
     }
```
Comment (designe) [inline euphony/build.gradle:94]: ### Here are my last suggestions :)

I added some build error handling code & modify some print messages for unit-test result.
```gradle
gradle.buildFinished {
        def cppBuildDir = ".cxx/Debug"
        def logfile = "android_gradle_build_stdout_targets.txt"
        def resultStr = "\n"
        def isUnitTestWorked = false
        def isDevice = false

        fileTree(cppBuildDir).matching {
            include "**/$logfile"
        }.forEach( file -> {
            if (file.exists()) {
                def lines = file.readLines()
                /* Check the unit-test working */
                if (lines.size() > 0) {
                    def lastSentence = lines.get(lines.size() - 1)
                    if (lastSentence.contains("**** Gtest Success ****")) {
                        isUnitTestWorked = true
                        isDevice = true
                    } else if (lastSentence.contains("ninja: no work to do.")) {
                        isUnitTestWorked = false
                        if (!lines.get(2).contains("0"))
                            isDevice = true
                    } else if (lastSentence.contains("Linking")) {
                        isUnitTestWorked = false
                        return
                    }

                    resultStr += "> $file.path\n"
                    resultStr += lines.join("\n")
                }
            }
        })

        print "\n"
        if (!isDevice) {
            println "THE BUILD WAS SUCCESSFUL, BUT NOT VERIFIED BY NATIVE UNIT TESTS."
            println "FOR UNIT TESTING, PLEASE CONNECT A VIRTUAL DEVICE OR ANDROID DEVICE TO ADB"
        }
        else {
            if (isUnitTestWorked) {
                println "Print build_stdout $resultStr"
                println "\nTHE BUILD & TESTS ARE SUCCESSFUL"
                println "THERE ARE NO ERRORS FOR NATIVE UNIT TESTS."
            }
            else {
                println "THERE ARE NO CHANGES TO THE CODE."
                println "THERE IS NO NEED FOR NATIVE UNIT TESTS."
            }
        }
    }
```

**Review by designe — COMMENTED** [review]
Here is my last comment!

File: euphony/build.gradle (hunk 12 above)
Comment (yeonns) [reply euphony/build.gradle:94]: Thank you for checking ```build.gradle```.
But, unfortunately, I got an error at below line: :cry: 
```
                        // error message; Index 2 out of bounds for length 2
                        if (!lines.get(2).contains("0"))

```

Comment (designe) [reply euphony/build.gradle:94]: Thanks for sharing printed line! :)

this is real last version I think :)

```groove
gradle.buildFinished {
    def cppBuildDir = ".cxx/Debug"
    def logfile = "android_gradle_build_stdout_targets.txt"
    def resultStr = "\n"
    def isUnitTestWorked = false
    def isDevice = false

    fileTree(cppBuildDir).matching {
        include "**/$logfile"
    }.forEach( file -> {
        if (file.exists()) {
            def lines = file.readLines()
            /* Check the unit-test working */
            if (lines.size() > 0) {
                def lastSentence = lines.get(lines.size() - 1)
                if (lastSentence.contains("**** Gtest Success ****")) {
                    isUnitTestWorked = true
                    isDevice = true
                } else if (lastSentence.contains("ninja: no work to do.")) {
                    isUnitTestWorked = false
                    if (lines.size() > 2 && !lines.get(2).contains("0"))
                        isDevice = true
                } else if (lastSentence.contains("Linking")) {
                    isUnitTestWorked = false
                    return
                }

                resultStr += "> $file.path\n"
                resultStr += lines.join("\n")
            }
        }
    })

    print "\n"
    if (!isDevice) {
        println "THE BUILD WAS SUCCESSFUL, BUT NOT VERIFIED BY NATIVE UNIT TESTS."
        println "FOR UNIT TESTING, PLEASE CONNECT A VIRTUAL DEVICE OR ANDROID DEVICE TO ADB"
    }
    else {
        if (isUnitTestWorked) {
            println "Print build_stdout $resultStr"
            println "\nTHE BUILD & TESTS ARE SUCCESSFUL"
            println "THERE ARE NO ERRORS FOR NATIVE UNIT TESTS."
        }
        else {
            println "THERE ARE NO CHANGES TO THE CODE."
            println "THERE IS NO NEED FOR NATIVE UNIT TESTS."
        }
    }
}
```

Comment (yeonns) [reply euphony/build.gradle]: @designe Thank you! I applied your last version :)

**Review by designe — APPROVED** [review]
big patch! 🦖 
Let me approve it!

**Pull request merged**